legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabProjectApi.java

```diff
@@ -143,7 +143,27 @@ else if (types instanceof Set)
                 {
                     stream = stream.filter(p -> p.getTagList().stream().anyMatch(tagSet::contains));
                 }
-                stream.map(p -> fromGitLabProject(p, mode)).forEach(projects::add);
+                // check limit
+                if (limit == null)
+                {
+                    stream.map(p -> fromGitLabProject(p, mode)).forEach(projects::add);
+                }
+                else
+                {
+                    // NOTE: this check implies that the mode that is scanned first could take all the slots within the
+                    // limit. This limitation hopefully will be removed when we remove support for prototype (UAT) mode.
+                    stream.map(p -> fromGitLabProject(p, mode)).limit(limit - projects.size()).forEach(projects::add);
+                    if (projects.size() > limit)
+                    {
+                        // If the number of projects found already exceed the limit, skip the check for the other modes
+                        break;
+                    }
+                }
+            }
+            // ensure the list of returned projects cannot exceed the limit (if specified) for whatever reasons
+            if (limit != null)
+            {
+                return projects.stream().limit(limit).collect(Collectors.toList());
```

Comment:
```suggestion
                while (projects.size() > limit)
                {
                    projects.remove(projects.size() - 1);
                };
```

Reply:
as discussed, now we use `subList` to not create an extra list. Thanks!